### PR TITLE
Type inference for macros

### PR DESCRIPTION
**Pull request by sheaf**
This PR implements a simple type-inference pass for macros,
see `HsBindgen.C.Tc.Macro`. It implements a type system with:

  - variables
  - function types
  - type constructors
  - quantified types

Type inference for a macro function definition proceeds in the
following stages:

  1. constraint generation following Algorithm W, creating new
     metavariables and carrying a substitution,
  2. constraint solving: constraint simplification and defaulting
  3. quantification over remaining (undefaulted) type variables
     and remaining (unsolved) constraints.

### CODE REVIEW

File: hs-bindgen/src/HsBindgen/C/Tc/Macro.hs [hunk 1]
```diff
@@ -0,0 +1,1121 @@
...
+      let !tvs' = goTy bound tvs argTy
+      in case NE.nonEmpty mbArgTys of
+        Nothing     -> goTy bound tvs resTy
+        Just argTys -> goFunTy bound tvs' argTys resTy
+
+tyVarSet :: [ TyVar ] -> IntSet
+tyVarSet = IntSet.fromList . map ( uniqueInt . tyVarUnique )
+
+newtype Subst tv = Subst { substMap :: IntMap ( tv, Type Ty ) }
+  deriving newtype Monoid
```
Comment (phadej) [inline hs-bindgen/src/HsBindgen/C/Tc/Macro.hs:345]: this makes `mappend` not equal to your `<>`.

File: hs-bindgen/src/HsBindgen/C/Tc/Macro.hs [hunk 2]
```diff
@@ -0,0 +1,1121 @@
...
+  | InstOrigin !FunName
+  | IntLitOrigin !( Literal Integer )
+  | FloatLitOrigin !Double
+  deriving stock Generic
+  deriving anyclass PrettyVal
+
+instance Show CtOrigin where
+  show = \case
+    AppOrigin fun ->
+      "In an application of '" ++ showFunName fun ++ "'."
```
Comment (phadej) [inline hs-bindgen/src/HsBindgen/C/Tc/Macro.hs:417]: I'm not a fan of non-derived like `Show` instances. ping @edsko

File: hs-bindgen/src/HsBindgen/Pretty/Orphans.hs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+--------------------------------------------------------------------------------
+
+deriving anyclass instance PrettyVal a => PrettyVal ( NE.NonEmpty a )
+
+instance PrettyVal a => PrettyVal ( Vec n a ) where
+  prettyVal v = Pretty.Con "Vec" [ prettyVal ( toList v ) ]
+
+instance PrettyVal a => PrettyVal ( IntMap a ) where
+  prettyVal v = Pretty.Con "IntMap" [ prettyVal ( toList v ) ]
```
Comment (phadej) [inline hs-bindgen/src/HsBindgen/Pretty/Orphans.hs:31]: this strips keys, is it intentional?

File: hs-bindgen/src/HsBindgen/C/Tc/Macro.hs (hunk 2 above)
Comment (sheaf) [reply hs-bindgen/src/HsBindgen/C/Tc/Macro.hs]: Fair point. I made these into separate functions and derived the Show instance.

File: hs-bindgen/src/HsBindgen/C/Tc/Macro.hs (hunk 1 above)
Comment (sheaf) [reply hs-bindgen/src/HsBindgen/C/Tc/Macro.hs]: Yes indeed, thanks. Fixed.

**Pull request merged**